pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -16,143 +16,602 @@
+  private final String _sortedColumn;
+  private Logger segmentLogger = LOGGER;
+  private final String _tableStreamName;
+  private AtomicLong _lastUpdatedRawDocuments = new AtomicLong(0);
+  private final String _instance;
+  private final ServerSegmentCompletionProtocolHandler _protocolHandler;
+
+
+  private boolean endCriteriaReached() {
+    Preconditions.checkState(_state.shouldConsume(), "Incorrect state %s", _state);
+    if (_state.equals(State.INITIAL_CONSUMING)) {
+      if (_consumeEndTime > 0) {
+        // We have a time-based constraint.
+        long now = now();
+        if (now >= _consumeEndTime) {
+          segmentLogger.info("Stopping consumption due to time limit start={} now={}", _startTimeMs, now);
+          return true;
+        }
+      } else {
+        // We have a number-of-rows based constraint.
+        if (_numRowsConsumed >= _segmentMaxRowCount) {
+          segmentLogger.info("Stopping consumption due to row limit nRows={} maxNRows={}", _numRowsConsumed,
+              _segmentMaxRowCount);
+          return true;
+        }
+      }
+      return false;
+    } else if (_state.equals(State.CATCHING_UP)) {
+      if (_currentOffset == _finalOffset) {
+        segmentLogger.info("Caught up to offset {}", _finalOffset);
+        return true;
+      }
+      if (_currentOffset > _finalOffset) {
+        throw new RuntimeException("Offset higher in catchup mode: current " + _currentOffset + " final " + _finalOffset);
+      }
+    }
+    throw new RuntimeException("Unreachable");
+  }
+
+  private void consumeLoop() {
+    while(!_receivedStop && !endCriteriaReached()) {
+      // Consume for the next _kafkaReadTime ms, or we get to final offset, whichever happens earlier,
+      // Update _currentOffset upon return from this method
+      Iterable<MessageAndOffset> messagesAndOffsets = _consumerWrapper.fetchMessages(_currentOffset, _endOffset, KAFKA_MAX_FETCH_TIME_MILLIS);
+      Iterator<MessageAndOffset> msgIterator = messagesAndOffsets.iterator();
+
+      int batchSize = 0;
+      while (!_receivedStop && !endCriteriaReached() && msgIterator.hasNext()) {
+        // Get a batch of messages from Kafka
+        // Index each message
+        MessageAndOffset messageAndOffset = msgIterator.next();
+        byte[] array = messageAndOffset.message().payload().array();
+        int offset = messageAndOffset.message().payload().arrayOffset();
+        int length = messageAndOffset.message().payloadSize();
+        GenericRow row = _messageDecoder.decode(array, offset, length);
+
+        if (row != null) {
+          row = _fieldExtractor.transform(row);
+          boolean canTakeMore = _realtimeSegment.index(row);  // Ignore the boolean return
+          if (!canTakeMore) {
+            segmentLogger.warn("We got full during indexing");
+          }
+          batchSize++;
+        }
+        _currentOffset = messageAndOffset.nextOffset();
+        _numRowsConsumed++;
```

Comment:
This assumes that each row that is consumed from Kafka is valid. We should only consider valid rows that were successfully indexed, not just any row.